From 887e86e97a4969b36fb4d95e9e0f238321fad856 Mon Sep 17 00:00:00 2001
From: Balamurugan Mahalingam <quic_bmahalin@quicinc.com>
Date: Thu, 4 Jul 2024 01:12:53 -0700
Subject: [PATCH] QSDK: sigma-dut: Add support to configure ppe_vp_type to test
 different datapath

Setup the different datapath options using the new command line option -X.
Valid values for the command line option 'X' are passive, active and ds.
Default option would be ds if no option is specified.

Option passive takes the data traffic through SFE (Shortcut Forwarding Engine).
Option active takes the data traffic through PPE_VP (HW assisted datapath).
Option passive takes the data traffic through DS (Direct Switching).

Patch-dependency: None
Patch-work: None

Signed-off-by: Balamurugan Mahalingam <quic_bmahalin@quicinc.com>
---
 ap.c        |  4 ++++
 sigma_dut.c | 16 +++++++++++++++-
 sigma_dut.h |  1 +
 3 files changed, 20 insertions(+), 1 deletion(-)

--- a/ap.c
+++ b/ap.c
@@ -9782,6 +9782,8 @@ write_conf:
 	else
 		fprintf(f, "ctrl_interface=/var/run/hostapd\n");
 
+	fprintf(f, "ppe_vp=%d\n", dut->ppe_vp_type);
+
 	if ((dut->ap_ssid[0] && conf_counter == 0) ||
 	    (dut->ap_ssid[0] && dut->ap_is_mld))
 		fprintf(f, "ssid=%s\n", dut->ap_ssid);
@@ -10835,6 +10837,7 @@ skip_vht_parameters_set:
 			else
 				fprintf(f, "ctrl_interface=/var/run/hostapd\n");
 			fprintf(f, "ssid=%s\n", dut->mbss.conf[i].ssid);
+			fprintf(f, "ppe_vp=%d\n", dut->ppe_vp_type);
 			if (dut->bridge)
 				fprintf(f, "bridge=%s\n", dut->bridge);
 
@@ -10935,6 +10938,7 @@ skip_vht_parameters_set:
 			else
 				fprintf(f, "ctrl_interface=/var/run/hostapd\n");
 			fprintf(f, "ssid=%s\n", dut->cohosted[dut->cohost_iface].conf[i].ssid);
+			fprintf(f, "ppe_vp=%d\n", dut->ppe_vp_type);
 			if (dut->bridge)
 				fprintf(f, "bridge=%s\n", dut->bridge);
 
--- a/sigma_dut.c
+++ b/sigma_dut.c
@@ -1025,6 +1025,7 @@ static void usage(void)
 	       "       [-N <device_get_info vendor>] \\\n"
 	       "       [-o <device_get_info model>] \\\n"
 	       "       [-O <device_get_info version>] \\\n"
+	       "       [-X <configure datapath mode. passive, active, ds> ] \\\n"
 #ifdef MIRACAST
 	       "       [-x <sink|source>] \\\n"
 	       "       [-y <Miracast library path>] \\\n"
@@ -1057,7 +1058,7 @@ int main(int argc, char *argv[])
 
 	for (;;) {
 		c = getopt(argc, argv,
-			   "aAb:Bc:C:dDE:e:fF:gGhH:j:J:i:Ik:K:l:L:m:M:nN:o:O:p:P:qQr:R:s:S:tT:uU:v:VWw:x:y:z:Z:2345:6:7");
+			   "aAb:Bc:C:dDE:e:fF:gGhH:j:J:i:Ik:K:l:L:m:M:nN:o:O:p:P:qQr:R:s:S:tT:uU:v:VWw:xX:y:z:Z:2345:6:7");
 		if (c < 0)
 			break;
 		switch (c) {
@@ -1236,6 +1237,19 @@ int main(int argc, char *argv[])
 		case 'A':
 			sigma_dut.sim_no_username = 1;
 			break;
+		case 'X':
+			if (strcmp(optarg, "passive") == 0) {
+				sigma_dut.ppe_vp_type = 1;
+			} else if (strcmp(optarg, "active") == 0) {
+				sigma_dut.ppe_vp_type = 2;
+			} else if (strcmp(optarg, "ds") == 0) {
+				sigma_dut.ppe_vp_type = 3;
+			} else {
+				sigma_dut.ppe_vp_type = 3;
+				sigma_dut_print(&sigma_dut, DUT_MSG_INFO,
+						"Override ppe_vp type to DS");
+			}
+			break;
 #ifdef MIRACAST
 		case 'x':
 			if (strcmp(optarg, "sink") == 0) {
--- a/sigma_dut.h
+++ b/sigma_dut.h
@@ -1267,6 +1267,7 @@ struct sigma_dut {
 	int btm_ml_link;
 #define MAX_MLD_SUPPORTED 2
 	enum ap_chwidth ap_link_chwidth[MAX_MLD_SUPPORTED];
+	int ppe_vp_type;
 };
 
 
